### ISSUE TITLE
JDK8 module should respect JsonInclude.NON_ABSENT when serializing Optionals

### ISSUE THREAD

**Issue by emersonf** [FasterXML/jackson-datatype-jdk8#8]
```java
@JsonInclude(com.fasterxml.jackson.annotation.JsonInclude.Include.NON_NULL)
public class Foo {
    private String bar;

    public Optional<String> getBar() {
        return Optional.ofNullable(bar);
    }
}
```
gets serialised as 

```json
{
    "bar": null
}
```
but should be 

```json
{}
```
since `NON_NULL` is specified.

**Comment by cowtowncoder** [comment]
This is the expected behavior, although not well documented before 2.6.
Only Java `null`s are excluded, but not referential values or wrappers that could be dereferenced as nulls or null equivalent.
If you want to exclude absent Optionals, use `JsonInclude.Include.NON_EMPTY` instead; it will exclude both nulls and "empty" referential types, Maps and Collections.

This behavior is clarified as the expected default for Jackson 2.6 for all referential types (`AtomicReference`, Guava optional type) etc. I understand that what you expect is a reasonable alternative expectation, so we need to make sure to try to document this well.

**Comment by emersonf** [comment]
If `NON_NULL` would be too coarse if it included `Optional`s, wouldn't using `NON_EMPTY` to exclude `Optional`s be too coarse because it includes collections?

There's a semantic difference between an empty scalar and an empty collection in JSON. And there's a semantic difference between an empty `Optional<String>` and an empty `List<String>` in Java. Since `Optional`s are constructs designed solely to avoid NPEs, isn't another level warranted, e.g. `NON_NULL_EQUIVALENT` or something?

**Comment by cowtowncoder** [comment]
Use of global exclusion is rather coarse regardless. But while adding more levels would be useful from user perspective, it is difficult to implement in the backend. Ability to check for emptiness is via `JsonSerializer.isEmpty()` (which is why it is easier to support than null check -- general-purpose null check via reference types turns out to be very difficult to do, unfortunately, which is big part of deciding between null/empty).
Even some scalar types have concept of emptiness, an obvious example being "" for Strings.

So I guess what I am saying is that the idea of a new level makes sense, but I am not sure of how that could be implemented yet.

There are some other work-arounds, for example per-property `@JsonInclude` (instead of, or in addition to, global setting). That is not very convenient, but has the benefit of working as is.

I have also been thinking of custom inclusion checks, in form of passing object that defines inclusion criteria for value. So far I had thought mostly of per-property annotation, but global setting would have its uses.

So... I am open to ideas, but as of yet do not see something that would seem feasible to implement unfortunately.

**Comment by emersonf** [comment]
I've had a quick look through the code. How about the following

* create a new level like `NON_NULL_EQUIVALENT` (it's a daft name, but just treat it as an example)
* in `BeanPropertyWriter`, either add `MARKER_FOR_NULL_EQUIVALENT` or multiple markers for the specific cases, like `MARKER_FOR_EMPTY_OPTIONAL`, ..
* in `BuildWriter`, set `value(s)ToSuppress` to the markers for the `NON_NULL_EQUIVALENT` case
* in `BeanPropertyWriter`, handle those cases

I'm not familiar enough with the code to know the implications, but that seems to use existing controls.

**Comment by cowtowncoder** [comment]
@emersonf Thank you for the ideas, this is useful. I agree in that reuse of existing hooks makes sense.

One problem is that main problem is more with how to recognize referential types, and detect absence/presence. It is not possible to have dependencies from `jackson-databind` to extension modules, and the system should be extensible enough to support referential types by add-on. 
For inclusion check, it is necessary to detect presence/absence, and method to call depends on datatype in question.
For JDK8 check has to be handled by this module, since core component still only require JDK6.

However, I think that it might be possible to solve this, given that this feature (https://github.com/FasterXML/jackson-databind/issues/698) was added in 2.6.0.
If so, handling could piggyback on 'isEmpty()' check functionality, so that:

1. A new enum is needed: I'd probably suggest `NON_NULL_OR_ABSENT` (or perhaps just `NON_ABSENT`)
2. Detection of a type being referential (issue 698) used to check whether enum applies
3. For referential type, assume handling same as `NON_EMPTY`, for other types, `NON_NULL`

**Comment by emersonf** [comment]
+1 for naming it `NON_ABSENT` ... empty collections aren't absent, referential types may be. Good name.

I hadn't thought of dragging in the kitchen sink to support reference types, but it looks like you may have that under control. 

re: treating them as NON_EMPTY, just be careful that `NON_ABSENT` correctly handles the `Optional<List>` case when an empty list is present, i.e. it should serialize `[]`.

**Comment by cowtowncoder** [comment]
@emersonf Good point on `Optional<List>`, that should be non-transitive.

**Comment by cowtowncoder** [comment]
As per databind change, can now add support here as well. Support turned out to be just couple of lines of code.

**Comment by cowtowncoder** [comment]
Now implemented for 2.6.0-rc2.

**Comment by emersonf** [comment]
Sweet! Thanks for the quick implementation. :+1:
